### PR TITLE
Support hooking instructions

### PR DESCRIPTION
**Pull request by talliberman**
Added `attach_instruction` to `Interceptor` to allow instrumentation of specific instructions.

Implemented using an `InstructionInvocationListener` which has only 1 callback, instead of the 2 required by `InvocationListener`.

### CODE REVIEW

**Comment by talliberman** [comment]
Hi @meme, 

I tried to figure out why the CI failed and noticed that `cargo test` seems to fail on the main branch as well.

**Comment by meme** [comment]
Firstly, it's not clear why all this extra code is needed if the only functional difference is having on_leave set to NULL. Is it possible to instead change the C API for `GumRustInvocationListener` to support both modes of operation? Then we can reuse the existing code path.

> I tried to figure out why the CI failed and noticed that cargo test seems to fail on the main branch as well.

Seems like Rust stable 1.63.0 introduced a change which breaks our pipeline with a new error for `reference to packed field is unaligned`. I'll fix this in `main` so you can rebase. It seems that there are some duplicated symbol errors for your PR as well, though (same names in `instruction_invocation_listener.c` as `invocation_listener.c`)

**Comment by talliberman** [comment]
Thanks for your reply.

>  Firstly, it's not clear why all this extra code is needed if the only functional difference is having on_leave set to NULL. Is it possible to instead change the C API for GumRustInvocationListener to support both modes of operation? Then we can reuse the existing code path.

It may be possible but I couldn't find an easy way to reuse the code. If you know how it can be done let me know and I could implement it differently. If we want to reuse the code in `invocation_listener.c` the main hurdle is that the flag which indicates whether we are hooking a function with on_enter/on_leave or a specific instruction needs to be passed to the function `gum_rust_invocation_listener_iface_init` so that it can set `iface->on_leave = NULL`.

I took a look at Frida's implementation in `gum/guminvocationlistener.c`. It seems they implemented two different `GumInvocationListener`s: `GumCallListener` and `GumProbeListener`. Which is quite similar to my implementation.

> Seems like Rust stable 1.63.0 introduced a change which breaks our pipeline with a new error for reference to packed field is unaligned. I'll fix this in main so you can rebase. It seems that there are some duplicated symbol errors for your PR as well, though (same names in instruction_invocation_listener.c as invocation_listener.c)

Great thanks, when the main branch is fixed I will rebase and fix any errors that I added.

**Comment by meme** [comment]
> It may be possible but I couldn't find an easy way to reuse the code

Apologies if I am misunderstanding, but is it not as simple as replacing `instruction_invocation_listener_transform` with:

```rust
pub(crate) fn instruction_invocation_listener_transform<I: InstructionInvocationListener>(
    invocation_listener: &mut I,
) -> *mut frida_gum_sys::GumInvocationListener {
    let rust = frida_gum_sys::RustInvocationListenerVTable {
        user_data: invocation_listener as *mut _ as *mut c_void,
        on_enter: Some(call_on_enter::<I>),
        on_leave: None, // <--- Use None instead of setting the 
    };

    unsafe { frida_gum_sys::gum_rust_invocation_listener_new(rust) }
}
```

Which makes `on_leave` NULL without duplicating the code?

**Comment by talliberman** [comment]
I assumed (because of [this](https://github.com/frida/frida-gum/issues/445)) that we must set `iface->on_leave = NULL` in `gum_rust_invocation_listener_iface_init`, but it seems like setting it to `gum_rust_invocation_listener_on_leave` and not calling the rust callback if it's `NULL` inside of `gum_rust_invocation_listener_on_leave` works as well.

I implemented it that way and pushed.

File: frida-gum-sys/invocation_listener.c [hunk 1]
```diff
@@ -39,7 +39,9 @@ static void
 gum_rust_invocation_listener_on_leave(GumInvocationListener *listener, GumInvocationContext *context)
 {
   GumRustInvocationListener *self = GUM_RUST_INVOCATION_LISTENER(listener);
-  return self->rust.on_leave(self->rust.user_data, context);
+  if (NULL != self->rust.on_leave){
```
Comment (meme) [inline frida-gum-sys/invocation_listener.c:42]:
```suggestion
  if (NULL != self->rust.on_leave) {
```

File: frida-gum/src/interceptor.rs [hunk 2]
```diff
@@ -61,6 +61,25 @@ impl<'a> Interceptor<'a> {
...
+    /// Attach a listener to an instruction address.
+    ///
+    /// # Safety
+    ///
+    /// The provided address *must* point to a valid instruction.
+    #[cfg(feature = "invocation-listener")]
+    #[cfg_attr(doc_cfg, doc(cfg(feature = "invocation-listener")))]
+    pub fn attach_instruction<I: ProbeListener>(
+        &mut self,
+        f: NativePointer,
```
Comment (meme) [inline frida-gum/src/interceptor.rs:73]:
```suggestion
        instr: NativePointer,
```

**Review by meme — CHANGES_REQUESTED** [review]
Few nits. Other than that LGTM. Thanks for the PR.

**Comment by oleavr** [comment]
I'm afraid the GObject interface `on_leave` must be `NULL`, otherwise `Interceptor` is going to replace the return address/LR and that will result in undefined behavior when the instruction selected isn't the first instruction of a function. So while it may appear to work, it's going to result in corruption in many cases.

**Comment by talliberman** [comment]
Thanks for your input @oleavr. Would you say that the implementation in 255a6bcfeee9b89059515d6cc18dea4e793e00a2 is OK?

Also, where is the code that skips the instrumentation of the return address/LR if `on_leave` is `NULL`?

**Comment by talliberman** [comment]
I brought back the implementation with a separate listener. I renamed `InstructionInvocationListener` -> `ProbeListener` as that is more consistent with the implementation in `Frida`.

File: frida-gum/src/interceptor.rs (hunk 2 above)
Comment (talliberman) [reply frida-gum/src/interceptor.rs]: Fixed

File: frida-gum-sys/build.rs [hunk 3]
```diff
@@ -108,6 +111,11 @@ fn main() {
             .file("invocation_listener.c")
             .opt_level(3)
             .compile("invocation_listener");
+
+            builder
```
Comment (meme) [inline frida-gum-sys/build.rs:115]:
```suggestion
        builder
```

**Review by meme — CHANGES_REQUESTED** [review]

**Comment by talliberman** [comment]
The failure in the CI is the same as #58

**Comment by talliberman** [comment]
I can't reproduce the issue from the CI, but I pushed an attempt to fix it.

**Review by meme — APPROVED** [review]
Sorry for the back-and-forth on this! LGTM.

**Pull request merged**